pkg/managed/clusterroles_test.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/verrazzano/verrazzano-operator/pkg/constants"
+	"github.com/verrazzano/verrazzano-operator/pkg/monitoring"
+	"github.com/verrazzano/verrazzano-operator/pkg/testutil"
+	"github.com/verrazzano/verrazzano-operator/pkg/types"
+	"github.com/verrazzano/verrazzano-operator/pkg/util"
+	rbacv1 "k8s.io/api/rbac/v1"
+	v1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"reflect"
+	"testing"
+)
+
+func TestCleanupOrphanedClusterRoles(t *testing.T) {
+	modelBindingPair := testutil.GetModelBindingPair()
+	clusterConnections := testutil.GetManagedClusterConnections()
+	clusterConnection := clusterConnections["cluster3"]
+
+	cr := v1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "orphaned",
+			Labels: map[string]string{
+				"verrazzano.binding": "testBinding",
+				"verrazzano.cluster": "cluster3",
+			},
+		},
+	}
+	_, err := clusterConnection.KubeClient.RbacV1().ClusterRoles().Create(context.TODO(), &cr, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("can't create cluster role: %v", err)
+	}
+
+	CleanupOrphanedClusterRoles(modelBindingPair, clusterConnections)
+
+	_, err = clusterConnection.KubeClient.RbacV1().ClusterRoles().Get(context.TODO(), "orphaned", metav1.GetOptions{})
+	if err == nil {
+		t.Fatal("expected cluster role to be cleaned up")
+	}
+}
+
+func TestCreateClusterRoles(t *testing.T) {
+	assert := assert.New(t)
+
+	modelBindingPair := testutil.GetModelBindingPair()
+	clusterConnections := testutil.GetManagedClusterConnections()
+	clusterConnection := clusterConnections["cluster1"]
+
+	CreateClusterRoles(modelBindingPair, clusterConnections)
+
+	list, err := clusterConnection.KubeClient.RbacV1().ClusterRoles().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("can't list cluster roles: %v", err)
+	}
+	assert.Equal(1, len(list.Items))
+	cr := list.Items[0]
+	assert.Equal("verrazzano-system", cr.Name)
+	assertRules(t, &cr, getRules())
+}
+
+func TestCreateClusterRolesVmiSystem(t *testing.T) {
+	assert := assert.New(t)
+
+	modelBindingPair := testutil.GetModelBindingPair()
+	clusterConnections := testutil.GetManagedClusterConnections()
+	clusterConnection := clusterConnections["cluster1"]
+
+	modelBindingPair.Binding.Name = constants.VmiSystemBindingName
+
+	CreateClusterRoles(modelBindingPair, clusterConnections)
+
+	list, err := clusterConnection.KubeClient.RbacV1().ClusterRoles().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("can't list cluster roles: %v", err)
+	}
+	expectedRoles := monitoring.GetSystemClusterRoles("cluster1")
+	assert.Equal(len(expectedRoles), len(list.Items))
+
+	for _, expected := range expectedRoles {
+		cr, err := clusterConnection.KubeClient.RbacV1().ClusterRoles().Get(context.TODO(), expected.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("can't get cluster role %s: %v", expected.Name, err)
+		}
+		assertRules(t, cr, expected.Rules)
+	}
+}
+
+func TestCreateClusterRolesUpdate(t *testing.T) {
```

Comment:
this test and the two before it seem to be very similar - can they be parameterized? i don't know the answer to that in go test, i know ginkgo can... probably needs a group discussion

Reply:
i don't understand the name of this test - what is is meant to do? i don't see any call to anything that looks like an update...?

Reply:
The three tests test different code paths through CreateClusterRoles.  I've updated the comments to make that clearer.  It was easier and I think cleaner in this case to write separate tests and not do a parameterized test.

The test is named Update because CreateClusterRoles actually does update an existing cluster role if it already exists.  I've updated the test name to be clearer (maybe?)  Maybe the function CreateClusterRoles should be renamed to CreateOrUpdateClusterRoles?  ... same pattern for every other resource that the operator manages.